### PR TITLE
Screen Warmth: Fix a bug with syncing screen warmth on start

### PR DESCRIPTION
**Pull request by zwim**
Today I got a strange behavior with AutoWarmth.

Starting the device (Sage) and going from NickelMenu to KOReader (during twilight), the warmth of the display did not change, although the correct warmth was displayed in the status bar.

A few reboots and KOReader restarts and Nickel-KOReader rounds later I found, that the first call of `Powerd:setWarmth` did nothing as `Powerd:frontlightWarmth()` (a somewhat ambiguous name; getWarmth would be better) returned the same value, as `setWarmth` wanted to set. 

So the question is, why did `frontlightWarmth()` return that value? Why to h.... does `frontlightWarmth()` knew, what value AutoWarmth wanted to set.

The PR fixes this special quirk by forcing `setWarmth()` to use the current value. That should not do much harm, as it gets called only on resume, AutoWarmth menu actions, leave standby and on midnight.

I am quite sure this PR does what it is expected to do. A hand full rounds of reboots, KOReader restarts and so on it works.

And now, as I wanted to reproduce the bug without the PR the strange news: _I can not reproduce it right now!_

Draft by now as either someone can confirm the bug or someone can explain the unwanted behaviour.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/10066)

### CODE REVIEW

**Comment by NiLuJe** [comment]
IIRC, on Kobo, the Device init does some pretty hairy gymnastics to workaround the lack of HW capabilities to *get* the frontlight level from the hardware (on < Mk.6 devices at least).

This *might* be somewhat related to that... somehow?

**Comment by NiLuJe** [comment]
TL;DR: That approach seems harmless enough, as I don't really see myself ever looking into the aforementioned mess unless I *really* have to ;p.

**Comment by zwim** [comment]
I saw the mess ;) and thought that an ad hoc fix here (in AutoWarmth) should not break anything on old devices.
If someone wants to use AutoWarmth this should do no harm.

A more sane approach to the issue would be to only call 
```
    Powerd:setWarmth(Powerd:frontlightWarmth(), true)
```
once on initialization of the plugin. 
As I can not reproduce the issue right now let's wait a few days, maybe things get clearer.

**Comment by NiLuJe** [comment]
Does this fare any better?

```diff
diff --git a/frontend/device/kobo/powerd.lua b/frontend/device/kobo/powerd.lua
index 582a82c9b..6f235a82a 100644
--- a/frontend/device/kobo/powerd.lua
+++ b/frontend/device/kobo/powerd.lua
@@ -170,7 +170,7 @@ function KoboPowerD:init()
         if self:isFrontlightOnHW() then
             -- On devices with a mixer, setIntensity will *only* set the FL, so, ensure we honor the warmth, too.
             if self.device:hasNaturalLightMixer() then
-               self:setWarmth(self.fl_warmth)
+               self:setWarmth(self.fl_warmth, true)
             end
             -- Use setIntensity to ensure it sets fl_intensity, and because we don't want the ramping behavior of turnOn
             self:setIntensity(self:frontlightIntensityHW())
@@ -375,13 +375,6 @@ function KoboPowerD:afterResume()
     if self.fl == nil then return end
     -- Don't bother if the light was already off on suspend
     if not self.fl_was_on then return end
-    -- Update warmth state
-    if self.fl_warmth ~= nil then
-        -- And we need an explicit setWarmth if the device has a mixer, because turnOn won't touch the warmth on those ;).
-        if self.device:hasNaturalLightMixer() then
-            self:setWarmth(self.fl_warmth)
-        end
-    end
     -- Turn the frontlight back on
     self:turnOnFrontlight()
 
```

The first chunk ensures we *actually* set the warmth on init (as otherwise, the dupe check in `setWarmth` ensures `setWarmth(self.fl_warmth)` will do *nothing*, as it checks if the arg is equal to `self.fl_warmth`, i.e., itself.

The second chunk removes a similarly useless leftover from the old  in-powerd autowarmth support.

**Comment by zwim** [comment]
Some interesting news:
A `grep Color Kobo\ eReader.conf` shows, that the `ColorSetting` string sometimes does not match the real fl_warmth. It can happen the `ColorSetting` string is on a low temperature even if the screen is cold blue. After some time the string matches the screen color.
1.  So KOReader does what it should, but it gets wrong infos from nickel conf. 

2. @NiLuJe your hint in https://github.com/koreader/koreader/pull/10066#issuecomment-1412782204 does not work even if it should. Strange behavior. With your suggestion I got weird behavior sometimes, that the warmth was right, but the intensity was at max. As you, I won't want into that depth of this. :(

Maybe `koreader.sh` should do some (more) `sync` before killing nickel? I will try that. Did not work :(

The good news is, that my proposed solution works (and does not do any harm).

**Comment by NiLuJe** [comment]
> Maybe `koreader.sh` should do some (more) `sync` before killing nickel? I will try that. Did not work :(

I understand with this comment that you suspect that the wrong `ColorSetting` in the config is simply Nickel's *actual* state not having been flushed to the config yet, right?

In which case, yeah, there's nothing we can do I mean, NM could theoretically find whatever's responsible from flushing the relevant QSettings to disk, but that's way out of scope and NM-specific ;p.



> 2\. With your suggestion I got weird behavior sometimes, that the warmth was right, but the intensity was at max.

By `intensity` you mean the actual frontlight brightness, right? That's... *extremely* weird, because that snipper doesn't affect brightness at all, ever.

**Comment by zwim** [comment]
Now I see clearer:
`self:setWarmth(self.fl_warmth, true)` (in line 175 in kobo/powerd.lua) uses the default warmth_scale and not the device specific scale, as the device gets initialized later :( 

On a Sage writing values > 10 to `self:setWarmth(self.fl_warmth, true)` gives funny results: changing the frontlight brightness, random warmth, segmentation fault (try `self:setWarmth(self.fl_warmth, true)` on the console).

This explains the problems in my last comment.

So we need to calculate a correct `warmth_scale` in (or before) `_syncKoboLightOnStart` to avoid the funny bahavior.

**Comment by NiLuJe** [comment]
Oh, duh', nice catch!

As a minor nit, I'd move it right next to the min/max getter, because it's the reason generic does it post-init ;).

e.g.,

```diff
diff --git a/frontend/device/kobo/powerd.lua b/frontend/device/kobo/powerd.lua
index 22b6dc6d5..1dd8f337c 100644
--- a/frontend/device/kobo/powerd.lua
+++ b/frontend/device/kobo/powerd.lua
@@ -27,9 +27,6 @@ function KoboPowerD:_syncKoboLightOnStart()
     local new_warmth = nil
     local kobo_light_on_start = tonumber(G_defaults:readSetting("KOBO_LIGHT_ON_START"))
 
-    -- calculate the warmth scale that fits the device
-    self.warmth_scale = 100 / self.fl_warmth_max
-
     if kobo_light_on_start then
         if kobo_light_on_start > 0 then
             new_intensity = math.min(kobo_light_on_start, 100)
@@ -150,6 +147,8 @@ function KoboPowerD:init()
             -- Does this device's NaturalLight use a custom scale?
             self.fl_warmth_min = self.device.frontlight_settings.nl_min or self.fl_warmth_min
             self.fl_warmth_max = self.device.frontlight_settings.nl_max or self.fl_warmth_max
+            -- Generic does it *after* init, but we're going to need it *now*...
+            self.warmth_scale = 100 / self.fl_warmth_max
             -- If this device has a mixer, we can use the ioctl for brightness control, as it's much lower latency.
             if self.device:hasNaturalLightMixer() then
                 local kobolight = require("ffi/kobolight")
```

**Review by NiLuJe — APPROVED** [review]
Thanks for figuring that one out ;).
